File: backend/services/stigmer-server/pkg/domain/workflowexecution/controller/execution_filter_test.go
package workflowexecution

import (
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
	"google.golang.org/protobuf/types/known/durationpb"
	"google.golang.org/protobuf/types/known/structpb"
	"google.golang.org/protobuf/types/known/timestamppb"

	workflowexecutionv1 "github.com/stigmer/stigmer/apis/stubs/go/ai/stigmer/agentic/workflowexecution/v1"
	apiresource "github.com/stigmer/stigmer/apis/stubs/go/ai/stigmer/commons/apiresource"
)

func makeExec(id string, phase workflowexecutionv1.ExecutionPhase, startedAt, completedAt string, costMicros int64, tasks []*workflowexecutionv1.WorkflowTask) *workflowexecutionv1.WorkflowExecution {
	return &workflowexecutionv1.WorkflowExecution{
		Metadata: &apiresource.ApiResourceMetadata{Id: id},
		Status: &workflowexecutionv1.WorkflowExecutionStatus{
			Phase:           phase,
			StartedAt:       startedAt,
			CompletedAt:     completedAt,
			TotalCostMicros: costMicros,
			Tasks:           tasks,
		},
	}
}

func ids(execs []*workflowexecutionv1.WorkflowExecution) []string {
	out := make([]string, len(execs))
	for i, e := range execs {
		out[i] = e.GetMetadata().GetId()
	}
	return out
}

func TestApplyFilterCriteria_NilFilter(t *testing.T) {
	execs := []*workflowexecutionv1.WorkflowExecution{
		makeExec("a", workflowexecutionv1.ExecutionPhase_EXECUTION_COMPLETED, "", "", 0, nil),
	}
	result := applyFilterCriteria(execs, nil)
	assert.Len(t, result, 1)
}

func TestApplyFilterCriteria_PhaseFilter(t *testing.T) {
	execs := []*workflowexecutionv1.WorkflowExecution{
		makeExec("completed", workflowexecutionv1.ExecutionPhase_EXECUTION_COMPLETED, "", "", 0, nil),
		makeExec("failed", workflowexecutionv1.ExecutionPhase_EXECUTION_FAILED, "", "", 0, nil),
		makeExec("running", workflowexecutionv1.ExecutionPhase_EXECUTION_IN_PROGRESS, "", "", 0, nil),
	}

	filter := &workflowexecutionv1.ExecutionFilterCriteria{
		Phases: []workflowexecutionv1.ExecutionPhase{
			workflowexecutionv1.ExecutionPhase_EXECUTION_FAILED,
		},
	}

	result := applyFilterCriteria(execs, filter)
	assert.Equal(t, []string{"failed"}, ids(result))
}

func TestApplyFilterCriteria_MultiplePhases(t *testing.T) {
	execs := []*workflowexecutionv1.WorkflowExecution{
		makeExec("completed", workflowexecutionv1.ExecutionPhase_EXECUTION_COMPLETED, "", "", 0, nil),
		makeExec("failed", workflowexecutionv1.ExecutionPhase_EXECUTION_FAILED, "", "", 0, nil),
		makeExec("running", workflowexecutionv1.ExecutionPhase_EXECUTION_IN_PROGRESS, "", "", 0, nil),
	}

	filter := &workflowexecutionv1.ExecutionFilterCriteria{
		Phases: []workflowexecutionv1.ExecutionPhase{
			workflowexecutionv1.ExecutionPhase_EXECUTION_COMPLETED,
			workflowexecutionv1.ExecutionPhase_EXECUTION_IN_PROGRESS,
		},
	}

	result := applyFilterCriteria(execs, filter)
	assert.Equal(t, []string{"completed", "running"}, ids(result))
}

func TestApplyFilterCriteria_StartedAfter(t *testing.T) {
	execs := []*workflowexecutionv1.WorkflowExecution{
		makeExec("old", workflowexecutionv1.ExecutionPhase_EXECUTION_COMPLETED, "2026-05-20T10:00:00Z", "2026-05-20T10:01:00Z", 0, nil),
		makeExec("new", workflowexecutionv1.ExecutionPhase_EXECUTION_COMPLETED, "2026-05-23T10:00:00Z", "2026-05-23T10:01:00Z", 0, nil),
	}

	after := parseTime("2026-05-22T00:00:00Z")
	filter := &workflowexecutionv1.ExecutionFilterCriteria{
		StartedAfter: timestamppb.New(after),
	}

	result := applyFilterCriteria(execs, filter)
	assert.Equal(t, []string{"new"}, ids(result))
}

func TestApplyFilterCriteria_DurationRange(t *testing.T) {
	execs := []*workflowexecutionv1.WorkflowExecution{
		makeExec("fast", workflowexecutionv1.ExecutionPhase_EXECUTION_COMPLETED, "2026-05-23T10:00:00Z", "2026-05-23T10:00:05Z", 0, nil),
		makeExec("slow", workflowexecutionv1.ExecutionPhase_EXECUTION_COMPLETED, "2026-05-23T10:00:00Z", "2026-05-23T10:05:00Z", 0, nil),
		makeExec("running", workflowexecutionv1.ExecutionPhase_EXECUTION_IN_PROGRESS, "2026-05-23T10:00:00Z", "", 0, nil),
	}

	filter := &workflowexecutionv1.ExecutionFilterCriteria{
		MaxDuration: durationpb.New(60_000_000_000), // 60 seconds
	}

	result := applyFilterCriteria(execs, filter)
	assert.Equal(t, []string{"fast"}, ids(result))
}

func TestApplyFilterCriteria_CostRange(t *testing.T) {
	execs := []*workflowexecutionv1.WorkflowExecution{
		makeExec("cheap", workflowexecutionv1.ExecutionPhase_EXECUTION_COMPLETED, "", "", 10_000, nil),
		makeExec("expensive", workflowexecutionv1.ExecutionPhase_EXECUTION_COMPLETED, "", "", 500_000, nil),
	}

	filter := &workflowexecutionv1.ExecutionFilterCriteria{
		MinCostMicros: 100_000,
	}

	result := applyFilterCriteria(execs, filter)
	assert.Equal(t, []string{"expensive"}, ids(result))
}

func TestApplyFilterCriteria_FailedTaskName(t *testing.T) {
	execs := []*workflowexecutionv1.WorkflowExecution{
		makeExec("a", workflowexecutionv1.ExecutionPhase_EXECUTION_FAILED, "", "", 0, []*workflowexecutionv1.WorkflowTask{
			{TaskName: "validate", Status: workflowexecutionv1.WorkflowTaskStatus_WORKFLOW_TASK_FAILED},
		}),
		makeExec("b", workflowexecutionv1.ExecutionPhase_EXECUTION_FAILED, "", "", 0, []*workflowexecutionv1.WorkflowTask{
			{TaskName: "send_email", Status: workflowexecutionv1.WorkflowTaskStatus_WORKFLOW_TASK_FAILED},
		}),
	}

	filter := &workflowexecutionv1.ExecutionFilterCriteria{
		FailedTaskName: "validate",
	}

	result := applyFilterCriteria(execs, filter)
	assert.Equal(t, []string{"a"}, ids(result))
}

func TestApplyFilterCriteria_HasRetries(t *testing.T) {
	retryMeta, err := structpb.NewStruct(map[string]interface{}{
		"retry_count": 3.0,
	})
	require.NoError(t, err)

	execs := []*workflowexecutionv1.WorkflowExecution{
		makeExec("retried", workflowexecutionv1.ExecutionPhase_EXECUTION_COMPLETED, "", "", 0, []*workflowexecutionv1.WorkflowTask{
			{TaskName: "flaky", Status: workflowexecutionv1.WorkflowTaskStatus_WORKFLOW_TASK_COMPLETED, Metadata: retryMeta},
		}),
		makeExec("clean", workflowexecutionv1.ExecutionPhase_EXECUTION_COMPLETED, "", "", 0, []*workflowexecutionv1.WorkflowTask{
			{TaskName: "stable", Status: workflowexecutionv1.WorkflowTaskStatus_WORKFLOW_TASK_COMPLETED},
		}),
	}

	filter := &workflowexecutionv1.ExecutionFilterCriteria{
		HasRetries: true,
	}

	result := applyFilterCriteria(execs, filter)
	assert.Equal(t, []string{"retried"}, ids(result))
}

func TestApplySortField_StartedAt(t *testing.T) {
	execs := []*workflowexecutionv1.WorkflowExecution{
		makeExec("b", workflowexecutionv1.ExecutionPhase_EXECUTION_COMPLETED, "2026-05-23T12:00:00Z", "", 0, nil),
		makeExec("a", workflowexecutionv1.ExecutionPhase_EXECUTION_COMPLETED, "2026-05-23T10:00:00Z", "", 0, nil),
		makeExec("c", workflowexecutionv1.ExecutionPhase_EXECUTION_COMPLETED, "2026-05-23T14:00:00Z", "", 0, nil),
	}

	applySortField(execs, workflowexecutionv1.ExecutionSortField_EXECUTION_SORT_FIELD_STARTED_AT, true)
	assert.Equal(t, []string{"a", "b", "c"}, ids(execs))

	applySortField(execs, workflowexecutionv1.ExecutionSortField_EXECUTION_SORT_FIELD_STARTED_AT, false)
	assert.Equal(t, []string{"c", "b", "a"}, ids(execs))
}

func TestApplySortField_Cost(t *testing.T) {
	execs := []*workflowexecutionv1.WorkflowExecution{
		makeExec("mid", workflowexecutionv1.ExecutionPhase_EXECUTION_COMPLETED, "", "", 200_000, nil),
		makeExec("low", workflowexecutionv1.ExecutionPhase_EXECUTION_COMPLETED, "", "", 50_000, nil),
		makeExec("high", workflowexecutionv1.ExecutionPhase_EXECUTION_COMPLETED, "", "", 500_000, nil),
	}

	applySortField(execs, workflowexecutionv1.ExecutionSortField_EXECUTION_SORT_FIELD_COST, true)
	assert.Equal(t, []string{"low", "mid", "high"}, ids(execs))
}

func TestApplyFilterCriteria_CombinedFilters(t *testing.T) {
	execs := []*workflowexecutionv1.WorkflowExecution{
		makeExec("match", workflowexecutionv1.ExecutionPhase_EXECUTION_FAILED, "2026-05-23T10:00:00Z", "2026-05-23T10:01:00Z", 200_000, []*workflowexecutionv1.WorkflowTask{
			{TaskName: "validate", Status: workflowexecutionv1.WorkflowTaskStatus_WORKFLOW_TASK_FAILED},
		}),
		makeExec("wrong-phase", workflowexecutionv1.ExecutionPhase_EXECUTION_COMPLETED, "2026-05-23T10:00:00Z", "2026-05-23T10:01:00Z", 200_000, nil),
		makeExec("wrong-cost", workflowexecutionv1.ExecutionPhase_EXECUTION_FAILED, "2026-05-23T10:00:00Z", "2026-05-23T10:01:00Z", 10, []*workflowexecutionv1.WorkflowTask{
			{TaskName: "validate", Status: workflowexecutionv1.WorkflowTaskStatus_WORKFLOW_TASK_FAILED},
		}),
	}

	filter := &workflowexecutionv1.ExecutionFilterCriteria{
		Phases:         []workflowexecutionv1.ExecutionPhase{workflowexecutionv1.ExecutionPhase_EXECUTION_FAILED},
		MinCostMicros:  100_000,
		FailedTaskName: "validate",
	}

	result := applyFilterCriteria(execs, filter)
	assert.Equal(t, []string{"match"}, ids(result))
}

File: backend/services/stigmer-server/src/domain/workflowexecution/__tests__/execution-filter.test.ts
/**
 * Pins the list filter/sort helpers against Go's execution_filter_test.go
 * case-for-case (11 cases), plus TS-side edges Go covers implicitly
 * through its type system: the strict RFC3339 gate (Go time.Parse rejects
 * date-only strings; naive Date.parse would accept them), the legacy
 * phase fallback, the descending comparator's tie stability, and the
 * camelCase retry key.
 */
import { create } from "@bufbuild/protobuf";
import type { MessageInitShape } from "@bufbuild/protobuf";
import { timestampFromMs } from "@bufbuild/protobuf/wkt";
import { describe, expect, it } from "vitest";

import { WorkflowExecutionSchema } from "@stigmer/protos/ai/stigmer/agentic/workflowexecution/v1/api_pb";
import type { WorkflowExecution } from "@stigmer/protos/ai/stigmer/agentic/workflowexecution/v1/api_pb";
import { WorkflowTaskSchema } from "@stigmer/protos/ai/stigmer/agentic/workflowexecution/v1/api_pb";
import {
  ExecutionPhase,
  WorkflowTaskStatus,
} from "@stigmer/protos/ai/stigmer/agentic/workflowexecution/v1/enum_pb";
import {
  ExecutionFilterCriteriaSchema,
  ExecutionSortField,
} from "@stigmer/protos/ai/stigmer/agentic/workflowexecution/v1/io_pb";

import {
  applyFilterCriteria,
  applyLegacyPhaseFilter,
  applySortField,
  executionDurationMs,
  extractRetryCountFromMetadata,
  parseRfc3339Ms,
} from "../execution-filter.js";

type TaskInit = MessageInitShape<typeof WorkflowTaskSchema>;

function makeExec(
  id: string,
  phase: ExecutionPhase,
  startedAt: string,
  completedAt: string,
  costMicros: bigint,
  tasks: TaskInit[] = [],
): WorkflowExecution {
  return create(WorkflowExecutionSchema, {
    metadata: { id },
    status: {
      phase,
      startedAt,
      completedAt,
      totalCostMicros: costMicros,
      tasks,
    },
  });
}

function ids(executions: WorkflowExecution[]): string[] {
  return executions.map((execution) => execution.metadata?.id ?? "");
}

describe("applyFilterCriteria (execution_filter_test.go case-for-case)", () => {
  it("nil filter passes everything through", () => {
    const execs = [
      makeExec("a", ExecutionPhase.EXECUTION_COMPLETED, "", "", 0n),
    ];
    expect(applyFilterCriteria(execs, undefined)).toHaveLength(1);
  });

  it("single phase filter", () => {
    const execs = [
      makeExec("completed", ExecutionPhase.EXECUTION_COMPLETED, "", "", 0n),
      makeExec("failed", ExecutionPhase.EXECUTION_FAILED, "", "", 0n),
      makeExec("running", ExecutionPhase.EXECUTION_IN_PROGRESS, "", "", 0n),
    ];
    const filter = create(ExecutionFilterCriteriaSchema, {
      phases: [ExecutionPhase.EXECUTION_FAILED],
    });
    expect(ids(applyFilterCriteria(execs, filter))).toEqual(["failed"]);
  });

  it("multiple phases filter", () => {
    const execs = [
      makeExec("completed", ExecutionPhase.EXECUTION_COMPLETED, "", "", 0n),
      makeExec("failed", ExecutionPhase.EXECUTION_FAILED, "", "", 0n),
      makeExec("running", ExecutionPhase.EXECUTION_IN_PROGRESS, "", "", 0n),
    ];
    const filter = create(ExecutionFilterCriteriaSchema, {
      phases: [
        ExecutionPhase.EXECUTION_COMPLETED,
        ExecutionPhase.EXECUTION_IN_PROGRESS,
      ],
    });
    expect(ids(applyFilterCriteria(execs, filter))).toEqual([
      "completed",
      "running",
    ]);
  });

  it("startedAfter excludes earlier starts and unparseable starts", () => {
    const execs = [
      makeExec(
        "old",
        ExecutionPhase.EXECUTION_COMPLETED,
        "2026-05-20T10:00:00Z",
        "2026-05-20T10:01:00Z",
        0n,
      ),
      makeExec(
        "new",
        ExecutionPhase.EXECUTION_COMPLETED,
        "2026-05-23T10:00:00Z",
        "2026-05-23T10:01:00Z",
        0n,
      ),
      makeExec("no-start", ExecutionPhase.EXECUTION_COMPLETED, "", "", 0n),
    ];
    const filter = create(ExecutionFilterCriteriaSchema, {
      startedAfter: timestampFromMs(Date.parse("2026-05-22T00:00:00Z")),
    });
    expect(ids(applyFilterCriteria(execs, filter))).toEqual(["new"]);
  });

  it("duration range excludes over-max and incomplete executions", () => {
    const execs = [
      makeExec(
        "fast",
        ExecutionPhase.EXECUTION_COMPLETED,
        "2026-05-23T10:00:00Z",
        "2026-05-23T10:00:05Z",
        0n,
      ),
      makeExec(
        "slow",
        ExecutionPhase.EXECUTION_COMPLETED,
        "2026-05-23T10:00:00Z",
        "2026-05-23T10:05:00Z",
        0n,
      ),
      makeExec(
        "running",
        ExecutionPhase.EXECUTION_IN_PROGRESS,
        "2026-05-23T10:00:00Z",
        "",
        0n,
      ),
    ];
    const filter = create(ExecutionFilterCriteriaSchema, {
      maxDuration: { seconds: 60n, nanos: 0 },
    });
    expect(ids(applyFilterCriteria(execs, filter))).toEqual(["fast"]);
  });

  it("min cost excludes cheaper executions", () => {
    const execs = [
      makeExec("cheap", ExecutionPhase.EXECUTION_COMPLETED, "", "", 10_000n),
      makeExec(
        "expensive",
        ExecutionPhase.EXECUTION_COMPLETED,
        "",
        "",
        500_000n,
      ),
    ];
    const filter = create(ExecutionFilterCriteriaSchema, {
      minCostMicros: 100_000n,
    });
    expect(ids(applyFilterCriteria(execs, filter))).toEqual(["expensive"]);
  });

  it("failedTaskName matches only failed tasks with that exact name", () => {
    const execs = [
      makeExec("a", ExecutionPhase.EXECUTION_FAILED, "", "", 0n, [
        {
          taskName: "validate",
          status: WorkflowTaskStatus.WORKFLOW_TASK_FAILED,
        },
      ]),
      makeExec("b", ExecutionPhase.EXECUTION_FAILED, "", "", 0n, [
        {
          taskName: "send_email",
          status: WorkflowTaskStatus.WORKFLOW_TASK_FAILED,
        },
      ]),
    ];
    const filter = create(ExecutionFilterCriteriaSchema, {
      failedTaskName: "validate",
    });
    expect(ids(applyFilterCriteria(execs, filter))).toEqual(["a"]);
  });

  it("hasRetries keeps only executions with a retried task", () => {
    const execs = [
      makeExec("retried", ExecutionPhase.EXECUTION_COMPLETED, "", "", 0n, [
        {
          taskName: "flaky",
          status: WorkflowTaskStatus.WORKFLOW_TASK_COMPLETED,
          metadata: { retry_count: 3 },
        },
      ]),
      makeExec("clean", ExecutionPhase.EXECUTION_COMPLETED, "", "", 0n, [
        {
          taskName: "stable",
          status: WorkflowTaskStatus.WORKFLOW_TASK_COMPLETED,
        },
      ]),
    ];
    const filter = create(ExecutionFilterCriteriaSchema, { hasRetries: true });
    expect(ids(applyFilterCriteria(execs, filter))).toEqual(["retried"]);
  });

  it("combined filters apply AND logic", () => {
    const failedValidate: TaskInit[] = [
      { taskName: "validate", status: WorkflowTaskStatus.WORKFLOW_TASK_FAILED },
    ];
    const execs = [
      makeExec(
        "match",
        ExecutionPhase.EXECUTION_FAILED,
        "2026-05-23T10:00:00Z",
        "2026-05-23T10:01:00Z",
        200_000n,
        failedValidate,
      ),
      makeExec(
        "wrong-phase",
        ExecutionPhase.EXECUTION_COMPLETED,
        "2026-05-23T10:00:00Z",
        "2026-05-23T10:01:00Z",
        200_000n,
      ),
      makeExec(
        "wrong-cost",
        ExecutionPhase.EXECUTION_FAILED,
        "2026-05-23T10:00:00Z",
        "2026-05-23T10:01:00Z",
        10n,
        failedValidate,
      ),
    ];
    const filter = create(ExecutionFilterCriteriaSchema, {
      phases: [ExecutionPhase.EXECUTION_FAILED],
      minCostMicros: 100_000n,
      failedTaskName: "validate",
    });
    expect(ids(applyFilterCriteria(execs, filter))).toEqual(["match"]);
  });
});

describe("applySortField (execution_filter_test.go case-for-case)", () => {
  it("sorts by started_at both directions", () => {
    const execs = [
      makeExec("b", ExecutionPhase.EXECUTION_COMPLETED, "2026-05-23T12:00:00Z", "", 0n),
      makeExec("a", ExecutionPhase.EXECUTION_COMPLETED, "2026-05-23T10:00:00Z", "", 0n),
      makeExec("c", ExecutionPhase.EXECUTION_COMPLETED, "2026-05-23T14:00:00Z", "", 0n),
    ];
    applySortField(execs, ExecutionSortField.STARTED_AT, true);
    expect(ids(execs)).toEqual(["a", "b", "c"]);
    applySortField(execs, ExecutionSortField.STARTED_AT, false);
    expect(ids(execs)).toEqual(["c", "b", "a"]);
  });

  it("sorts by cost ascending", () => {
    const execs = [
      makeExec("mid", ExecutionPhase.EXECUTION_COMPLETED, "", "", 200_000n),
      makeExec("low", ExecutionPhase.EXECUTION_COMPLETED, "", "", 50_000n),
      makeExec("high", ExecutionPhase.EXECUTION_COMPLETED, "", "", 500_000n),
    ];
    applySortField(execs, ExecutionSortField.COST, true);
    expect(ids(execs)).toEqual(["low", "mid", "high"]);
  });

  it("sorts by duration with incomplete executions tied at -1 first", () => {
    const execs = [
      makeExec(
        "long",
        ExecutionPhase.EXECUTION_COMPLETED,
        "2026-05-23T10:00:00Z",
        "2026-05-23T10:10:00Z",
        0n,
      ),
      makeExec("incomplete-1", ExecutionPhase.EXECUTION_IN_PROGRESS, "2026-05-23T10:00:00Z", "", 0n),
      makeExec(
        "short",
        ExecutionPhase.EXECUTION_COMPLETED,
        "2026-05-23T10:00:00Z",
        "2026-05-23T10:00:01Z",
        0n,
      ),
      makeExec("incomplete-2", ExecutionPhase.EXECUTION_IN_PROGRESS, "", "", 0n),
    ];
    applySortField(execs, ExecutionSortField.DURATION, true);
    // -1 duration ties keep scan order (stable sort; Go SliceStable).
    expect(ids(execs)).toEqual([
      "incomplete-1",
      "incomplete-2",
      "short",
      "long",
    ]);
  });

  it("sorts by status (phase enum value) and keeps ties stable descending", () => {
    const execs = [
      makeExec("failed-1", ExecutionPhase.EXECUTION_FAILED, "", "", 0n),
      makeExec("pending", ExecutionPhase.EXECUTION_PENDING, "", "", 0n),
      makeExec("failed-2", ExecutionPhase.EXECUTION_FAILED, "", "", 0n),
    ];
    applySortField(execs, ExecutionSortField.STATUS, false);
    // Descending is Go's strictly-greater-first: FAILED(4) > PENDING(1);
    // the two FAILED entries tie and keep their scan order.
    expect(ids(execs)).toEqual(["failed-1", "failed-2", "pending"]);
  });

  it("unparseable started_at sorts like Go's zero time (before real instants)", () => {
    const execs = [
      makeExec("real", ExecutionPhase.EXECUTION_COMPLETED, "2026-05-23T10:00:00Z", "", 0n),
      makeExec("empty", ExecutionPhase.EXECUTION_COMPLETED, "", "", 0n),
    ];
    applySortField(execs, ExecutionSortField.STARTED_AT, true);
    expect(ids(execs)).toEqual(["empty", "real"]);
  });
});

describe("legacy phase filter and helpers", () => {
  it("applyLegacyPhaseFilter is a no-op for UNSPECIFIED", () => {
    const execs = [
      makeExec("a", ExecutionPhase.EXECUTION_COMPLETED, "", "", 0n),
      makeExec("b", ExecutionPhase.EXECUTION_FAILED, "", "", 0n),
    ];
    expect(
      applyLegacyPhaseFilter(execs, ExecutionPhase.EXECUTION_PHASE_UNSPECIFIED),
    ).toHaveLength(2);
    expect(
      ids(applyLegacyPhaseFilter(execs, ExecutionPhase.EXECUTION_FAILED)),
    ).toEqual(["b"]);
  });

  it("extractRetryCountFromMetadata reads both key spellings and rejects non-numbers", () => {
    const snake = create(WorkflowExecutionSchema, {
      status: { tasks: [{ metadata: { retry_count: 2 } }] },
    });
    const camel = create(WorkflowExecutionSchema, {
      status: { tasks: [{ metadata: { retryCount: 1 } }] },
    });
    const stringValue = create(WorkflowExecutionSchema, {
      status: { tasks: [{ metadata: { retry_count: "3" } }] },
    });
    expect(
      extractRetryCountFromMetadata(snake.status!.tasks[0]),
    ).toBe(2);
    expect(extractRetryCountFromMetadata(camel.status!.tasks[0])).toBe(1);
    // Go GetNumberValue answers 0 for non-number kinds.
    expect(
      extractRetryCountFromMetadata(stringValue.status!.tasks[0]),
    ).toBe(0);
  });

  it("parseRfc3339Ms enforces Go time.Parse strictness", () => {
    expect(parseRfc3339Ms("2026-05-23T10:00:00Z")).toBe(
      Date.parse("2026-05-23T10:00:00Z"),
    );
    expect(parseRfc3339Ms("2026-05-23T10:00:00.123456789+05:30")).toBe(
      Date.parse("2026-05-23T10:00:00.123456789+05:30"),
    );
    // Go rejects all of these (zero time); a bare Date.parse would accept
    // the first two.
    expect(parseRfc3339Ms("2026-05-23")).toBeNaN();
    expect(parseRfc3339Ms("2026-05-23T10:00:00")).toBeNaN();
    expect(parseRfc3339Ms("")).toBeNaN();
    expect(parseRfc3339Ms("not a time")).toBeNaN();
  });

  it("executionDurationMs answers -1 without both timestamps", () => {
    expect(
      executionDurationMs(
        makeExec("x", ExecutionPhase.EXECUTION_IN_PROGRESS, "2026-05-23T10:00:00Z", "", 0n),
      ),
    ).toBe(-1);
    expect(
      executionDurationMs(
        makeExec(
          "y",
          ExecutionPhase.EXECUTION_COMPLETED,
          "2026-05-23T10:00:00Z",
          "2026-05-23T10:00:05Z",
          0n,
        ),
      ),
    ).toBe(5000);
  });
});
